### ISSUE TITLE
ant `compile` target should not depend on `clean` and `generate` target.

### ISSUE THREAD

**Issue by lemmy** [tlaplus/tlaplus#969]
The `compile` target should not depend on the `clean` and `generate` target to ensure that each target can be run independently.  
 For example, one might want to run `compile` without cleaning test classes, or regenerating the parser classes. Instead, `customBuild.xml` defines "convenience" targets at the top.

**Comment by ahelwer** [comment]
I believe compile should depend on the generate target, because it is an actual dependency of compile; the things done in generate need to be done before the things done in compile. If you break this dependency, along with your suggested changes to remove the grammar/code sync check in the CI, this guarantees the files will drift out of sync over time. I can tell you this from my long experience dealing with checked-in generated code with the tree-sitter grammar.

**Comment by ahelwer** [comment]
I will reply here to [your comment](https://github.com/tlaplus/tlaplus/pull/968#pullrequestreview-2133286764), since you locked the thread there. Here is the comment (emphasis mine):

>This check should be performed as part of the Ant build to a) ensure platform independence and b) to be enforced locally (pre-push). **However, given the stability of the language grammar, it can likely be skipped/removed altogether.**

Here is [a quote from you](https://github.com/tlaplus/tlaplus/issues/596#issuecomment-803623702) last year, on a bug I reported with the parser (emphasis mine):
> Given that there exists an easy workaround **and the fact that changing SANY is a major undertaking (manually modified javacc source files)**, this has very low priority.

The reason the grammar was not often changed was *because* it was so hard to change. If changing it was a major undertaking, then nobody would change it. Recent work changes this dynamic; it:
1. Re-established the JavaCC grammar as the source of truth in PR #875
2. Added a *huge* number of tests for the JavaCC-generated parser in PRs #866 and #896
3. Ran JavaCC during the Ant build & CI to ensure the work in (1) does not face regression in PR #876 

Now, it is easy to change the grammar. So, the grammar will change more often. Reverting this work because the grammar was never change ignores that this work *specifically addressed* the reasons the grammar was never changed.

**Comment by lemmy** [comment]
https://github.com/tlaplus/tlaplus/issues/969#issuecomment-2183339780 is orthogonal to whether the `compile` target should rely on the `clean` and `generate` targets.  Clearly, compiling will fail if the `generate` target hasn't been executed beforehand. However, subsequent compilations will succeed without needing to rerun the `generate` target.  `generate` and `info` should be added to the `default` target.

**Comment by lemmy** [comment]
> The reason the grammar was not often changed was because it was so hard to change.

The reason the grammar was not often changed was because a) there were no significant bugs, and b) TLA+ is a stable and mature language where changes are extremely rare.  The last addition was the introduction of the proof language approximately 13 years ago.

According to the charter of the TLA+ Foundation, the SLC is in charge of evolving the language.

**Comment by ahelwer** [comment]
There are [quite a few known bugs in SANY](https://github.com/tlaplus/tlaplus/issues?q=is%3Aissue+is%3Aopen+label%3ASANY). Anyway I think it is acceptable to remove `generate` as a dependency of `compile` *as long as* the CI check stays - the CI check is more important to ensuring the JavaCC grammar and generated sources remain in sync.

**Comment by Calvin-L** [comment]
> For example, one might want to run compile without cleaning test classes, or regenerating the parser classes. Instead, customBuild.xml defines "convenience" targets at the top.

TL;DR: I'm sympathetic to this use case, but I would propose that the convenience tasks should be the ones with obvious names like "compile", "test", and "dist"; the primitive dependency-free tasks should have longer names (e.g. "only-compile", "only-test", "only-dist", etc), since I expect the primitive tasks will be used a lot less often.

----

Using primitive tasks requires mentally tracking the state of the build system, and (I find) it is _very_ easy to waste time on mistakes that my computer could have prevented. Here's my favorite one, which becomes possible if we remove `clean` from the dependencies of `compile`:

```
$ git diff
diff --git a/tlatools/org.lamport.tlatools/customBuild.xml b/tlatools/org.lamport.tlatools/customBuild.xml
index 67a6e339e..e7fe9d8ee 100644
--- a/tlatools/org.lamport.tlatools/customBuild.xml
+++ b/tlatools/org.lamport.tlatools/customBuild.xml
@@ -135,7 +135,7 @@
        </target>
 
        <!-- Compiles the TLA+ tools code -->
-       <target name="compile" depends="clean, generate">
+       <target name="compile" depends="generate">
                <mkdir dir="${class.dir}" />
                <echo>
```
```
$ ant -f customBuild.xml compile
[...]
BUILD SUCCESSFUL
Total time: 4 seconds
```
```
$ rm -rf src/tlc2/value
```
```
$ ant -f customBuild.xml compile
[...]
BUILD SUCCESSFUL                         # !?!?!?
Total time: 1 second
```

Sure looks like we don't need any of the `tlc2.value.*` classes! (In reality, the compiled .class files have been left behind in `./class/tlc2/value`, and are discovered there by javac.)

(I think that part of the problem here is Ant's simplicity; it unconditionally re-does work like `generate` even when that is obviously unnecessary. Even good old Gnu Make does not have such a deficiency. As a result, advanced users can save a little time by surgically running exactly the necessary Ant tasks at any given time, when I think virtually everyone would rather let the computer track all the tasks that need doing to achieve a particular result.)

**Comment by Calvin-L** [comment]
(Sorry, closed the issue by mistake. I've re-opened it.)

**Comment by lemmy** [comment]
Targets can be renamed using a specific naming pattern to differentiate between those that are dependency-free and those that are convenient. The only reason this hasn't been done yet is that the current names are widely used in many places.
